Add bfloat16 support detection for MPS in is_torch_bf16_gpu_available()

andrerom opened the pull request:
Add `bfloat16` support detection for MPS (Apple Silicon) in `is_torch_bf16_gpu_available()`

# What does this PR do?

Makes sure to allow it and not throw on `bf16` usage with "Your setup doesn't support bf16/gpu." from `TrainingArguments`.

### Why?
`bfloat16` has been supported for a few years now in Metal [1](https://www.reddit.com/r/LocalLLaMA/comments/14pz4v0/apples_metal_is_getting_bfloat16_support/) and torch.mps [1](https://github.com/pytorch/pytorch/issues/150121) [2](https://github.com/pytorch/pytorch/issues/141864).

**IMPORTANT**! On Apple M1 and M2 bfloat16 is emulated in software _(by Apple's Metal framework)_ using float32 for hardware instead, meaning there is no performance benefit, so on a M1 or M2 you should rather use float16 or float32.

andrerom committed:
Add bfloat16 support detection for MPS (Apple Silicon) in is_torch_bf16_gpu_available

bfloat16 seems to have been supported for a few years now in Metal and torch.mps.

Make sure to allow it and not throw on bf16 usage with "Your setup doesn't support bf16/gpu." from TrainingArguments.

SunMarc commented on src/transformers/utils/import_utils.py:
```diff
@@ -636,6 +636,10 @@ def is_torch_bf16_gpu_available() -> bool:
         return True
     if is_torch_npu_available():
         return torch.npu.is_bf16_supported()
+    if is_torch_mps_available():
+        # There is no torch.mps.is_bf16_supported(), could return true and hope for the best here, but not all intel setups support this.
+        import platform
+        return platform.machine().lower() == "arm64"
```
did you check if bf16 actually works with bf16 ?

SunMarc commented on src/transformers/utils/import_utils.py:
```diff
@@ -636,6 +636,10 @@ def is_torch_bf16_gpu_available() -> bool:
         return True
     if is_torch_npu_available():
         return torch.npu.is_bf16_supported()
+    if is_torch_mps_available():
+        # There is no torch.mps.is_bf16_supported(), could return true and hope for the best here, but not all intel setups support this.
+        import platform
+        return platform.machine().lower() == "arm64"
```
did you check if bf16 actually works with bf16 ?

andrerom replied:
If I checked if bf16 works with MPS on Apple silicon?

Partly, I also took the info from Apple (Metal) and Torch to indicate it should work, any specific test I can run here locally?

SunMarc commented:
I have this PR in accelerate that enables this, can you update the PR wrt to this PR ? https://github.com/huggingface/accelerate/pull/3373

andrerom committed:
Check bf16 support for MPS using torch method

Actually seems method exists: https://github.com/pytorch/pytorch/blob/5859edf1130cec5a021ace5d5b0e18144808f757/torch/_dynamo/device_interface.py#L519

It simply checks if you are on MacOs 14 or higher.

andrerom committed:
Document Metal emulation for bf16 support

Add note about Metal emulation for bf16 support on M1/M2.

SunMarc commented:
oh this is nice but let's use `torch.backends.mps.is_macos_or_newer(14, 0) `instead. [torch.mps.is_bf16_supported()](https://github.com/pytorch/pytorch/blob/5859edf1130cec5a021ace5d5b0e18144808f757/torch/_dynamo/device_interface.py#L519) was added only 8 months ago which mean that it was probably part of the 2.8 release but the bf16 support should already there since 2.6 release.

andrerom committed:
Update bf16 support check for MPS backend

is_bf16_supported() not exposed even if defined on MPSInterface, use same approach as in accelerate pr.

andrerom commented:
@SunMarc PR her now takes same approach 👍 

And actually seems is_bf16_supported() isn't fully exposed on 2.8 even, at least I had issues with it here locally.

SunMarc reviewed (approved):
Nice ! Let's get this merged, I'll add fp16 compatibility in another PR

SunMarc committed:
Merge branch 'main' into patch-1
